In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/dota-2-matches/players.csv
/kaggle/input/dota-2-matches/ability_upgrades.csv
/kaggle/input/dota-2-matches/test_player.csv
/kaggle/input/dota-2-matches/match_outcomes.csv
/kaggle/input/dota-2-matches/teamfights_players.csv
/kaggle/input/dota-2-matches/chat.csv
/kaggle/input/dota-2-matches/purchase_log.csv
/kaggle/input/dota-2-matches/test_labels.csv
/kaggle/input/dota-2-matches/ability_ids.csv
/kaggle/input/dota-2-matches/item_ids.csv
/kaggle/input/dota-2-matches/player_time.csv
/kaggle/input/dota-2-matches/patch_dates.csv
/kaggle/input/dota-2-matches/yasp_sample.json
/kaggle/input/dota-2-matches/match.csv
/kaggle/input/dota-2-matches/player_ratings.csv
/kaggle/input/dota-2-matches/hero_names.csv
/kaggle/input/dota-2-matches/cluster_regions.csv
/kaggle/input/dota-2-matches/teamfights.csv
/kaggle/input/dota-2-matches/objectives.csv


The first question I want to answer is: Can we predict the match outcome based on the aggregates of heroes in the match?

To start we are going to read in the data, and start to combine the CSVs to get information about a single match. 

In [2]:
#Each match has one entry, in this case we are focused on match_id, and radiant_win only as our target.
columnsWeWant = ['match_id','radiant_win']
match = pd.read_csv('/kaggle/input/dota-2-matches/match.csv')
match = match[columnsWeWant]

#We are only intrested in this case on the hero_id and match_id, and player_slot. Note that: Player_slot: 0-4 are Radiant, 128-132 are Dire
columnsWeWant = ['match_id','hero_id','player_slot']
players = pd.read_csv('/kaggle/input/dota-2-matches/players.csv')
players = players[columnsWeWant]

#Change from players_slot to boolean for radiant team or not
players["radiant"] = (players['player_slot'] < 5)

#Groups by match id and radiant boolean, and for each will combine hero_id into a list
combinedHeroID = players.groupby(['match_id','radiant']).hero_id.apply(lambda x: x.tolist())

#We unstack so that we can get the team lists for each match_id into columns instead of a row for each match_id
combinedHeroID = combinedHeroID.unstack()

#We rename the columns to make more sense
combinedHeroID.columns = ['dire_heros','radiant_heros']

#Here we merge the two DataFrames together on the match_id columns, and make the match_id the index
merged = combinedHeroID.merge(match,left_on='match_id',right_on='match_id')
merged = merged.set_index('match_id')

merged

dire_heros          radiant_heros  radiant_win
match_id                                                            
0          [106, 102, 46, 7, 73]   [86, 51, 83, 11, 67]         True
1           [73, 22, 5, 67, 106]    [7, 82, 71, 39, 21]        False
2            [38, 7, 10, 12, 85]   [51, 109, 9, 41, 27]        False
3           [78, 19, 31, 40, 47]   [50, 44, 32, 26, 39]        False
4         [101, 100, 22, 67, 21]    [8, 39, 55, 87, 69]         True
...                          ...                    ...          ...
49995      [32, 7, 109, 35, 112]   [73, 86, 21, 20, 14]         True
49996       [36, 1, 112, 60, 71]  [93, 74, 100, 32, 85]         True
49997       [28, 102, 21, 9, 23]  [100, 68, 75, 39, 44]         True
49998         [46, 7, 29, 44, 3]    [56, 50, 2, 72, 30]         True
49999       [100, 9, 90, 73, 53]   [94, 19, 68, 35, 21]        False

[50000 rows x 3 columns]

In [3]:
#To make features for the machine learning model we split the list of heros into different columns for each one
merged[["r1", "r2", "r3", "r4", "r5"]] = merged["radiant_heros"].tolist()
merged[["d1", "d2", "d3", "d4", "d5"]] = merged["dire_heros"].tolist()

merged = merged.drop(["radiant_heros","dire_heros"],axis=1)
merged

radiant_win   r1   r2   r3  r4  r5   d1   d2   d3  d4   d5
match_id                                                            
0                True   86   51   83  11  67  106  102   46   7   73
1               False    7   82   71  39  21   73   22    5  67  106
2               False   51  109    9  41  27   38    7   10  12   85
3               False   50   44   32  26  39   78   19   31  40   47
4                True    8   39   55  87  69  101  100   22  67   21
...               ...  ...  ...  ...  ..  ..  ...  ...  ...  ..  ...
49995            True   73   86   21  20  14   32    7  109  35  112
49996            True   93   74  100  32  85   36    1  112  60   71
49997            True  100   68   75  39  44   28  102   21   9   23
49998            True   56   50    2  72  30   46    7   29  44    3
49999           False   94   19   68  35  21  100    9   90  73   53

[50000 rows x 11 columns]

In [4]:
#Here we start making the Machine Learning Model, we will start with a Random Forest Model

#Start by setting the prediction target
y = merged.radiant_win

#Next we set the features we are going to use
data_features = ["r1", "r2", "r3", "r4", "r5","d1", "d2", "d3", "d4", "d5"]
X = merged[data_features]

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics

#Here we split the data into training and validation data
train_X, val_X, train_y, val_y = train_test_split(X, y,random_state = 0)

#Make and fit a Random Forest Classifier Model using training data
forest_model = RandomForestClassifier(random_state=1)
forest_model.fit(train_X, train_y)

#Use the Model to predict the validation data and get a accuracy score based on that
preds = forest_model.predict(val_X)
print("Accuracy of the model is: ", metrics.accuracy_score(val_y, preds))

Accuracy of the model is:  0.51392


In [5]:
#Lets try out the model on a random game I got from OPENDOTA: https://www.opendota.com/  Make sure that we do it from the same patches
#https://www.opendota.com/matches/17955123, In this game Radiant are the winners
random_game = pd.DataFrame([(31, 32, 29, 39, 52, 9, 79, 28, 21, 33)],columns=data_features)
print(random_game)
print(forest_model.predict(random_game))

   r1  r2  r3  r4  r5  d1  d2  d3  d4  d5
0  31  32  29  39  52   9  79  28  21  33
[ True]


To test our Random Forest Model we looked up a random match inside the patch, and the model thinks that Radiant will win, In this match Radiant does win.

But if we look at our accuracy score we see that we are correct 51.392% of the time, this is barely better than a coinflip.

There are a few ways we can improve our model, lets try some of them below!

In [6]:
#One of the options we have to make our model better is to make use of more variables. We can use player_ratings, purchase log, ability_upgrade, tower and barrack statuses. Gpm, Xpm, Kills and Deaths.

#Each match has one entry, in this case we are focused on match_id, and radiant_win only as our target.
columnsWeWant = ['match_id','radiant_win','duration','tower_status_radiant','tower_status_dire','barracks_status_dire','barracks_status_radiant']
match = pd.read_csv('/kaggle/input/dota-2-matches/match.csv')
match = match[columnsWeWant]

#
columnsWeWant = ['match_id','hero_id','player_slot','gold_per_min','xp_per_min','kills','deaths']
players = pd.read_csv('/kaggle/input/dota-2-matches/players.csv')
players = players[columnsWeWant]
players

#Important for the tower and barrack status. They are Bitmask that represents a binary of which ones are standing. Ex: 63 would mean all barracks are still standing.

match_id  hero_id  player_slot  gold_per_min  xp_per_min  kills  \
0              0       86            0           347         362      9   
1              0       51            1           494         659     13   
2              0       83            2           350         385      0   
3              0       11            3           599         605      8   
4              0       67            4           613         762     20   
...          ...      ...          ...           ...         ...    ...   
499995     49999      100          128           468         626     16   
499996     49999        9          129           507         607     12   
499997     49999       90          130           371         404      5   
499998     49999       73          131           780         703      8   
499999     49999       53          132           520         533      6   

        deaths  
0            3  
1            3  
2            4  
3            4  
4            3  
...        ...  
499995       9  
499996       6  
499997       3  
499998       6  
499999       6  

[500000 rows x 7 columns]

In [7]:
#As usual we start by combining the match data and player data

#Change from players_slot to boolean for radiant team or not
players["radiant"] = (players['player_slot'] < 5)

#Groups by match id and radiant boolean, and for each will combine hero_id into a list
combinedPlayerData = players.groupby(['match_id','radiant'])

combinedHeroID = combinedPlayerData.hero_id.apply(lambda x: x.tolist())
combinedGPM = combinedPlayerData.gold_per_min.apply(lambda x: x.tolist())
combinedXPM = combinedPlayerData.xp_per_min.apply(lambda x: x.tolist())
combinedKills = combinedPlayerData.kills.apply(lambda x: x.tolist())
combinedDeaths = combinedPlayerData.deaths.apply(lambda x: x.tolist())

#We unstack so that we can get the team lists for each match_id into columns instead of a row for each match_id
combinedHeroID = combinedHeroID.unstack()
combinedGPM = combinedGPM.unstack()
combinedXPM = combinedXPM.unstack()
combinedKills = combinedKills.unstack()
combinedDeaths = combinedDeaths.unstack()

#We rename the columns to make more sense
combinedHeroID.columns = ['dire_heros','radiant_heros']
combinedGPM.columns = ['dire_heros_gpm','radiant_heros_gpm']
combinedXPM.columns = ['dire_heros_xpm','radiant_heros_xpm']
combinedKills.columns = ['dire_heros_kills','radiant_heros_kills']
combinedDeaths.columns = ['dire_heros_deaths','radiant_heros_deaths']

mergedPlayerData = combinedHeroID.merge(combinedGPM,left_on='match_id',right_on='match_id')
mergedPlayerData = mergedPlayerData.merge(combinedXPM,left_on='match_id',right_on='match_id')
mergedPlayerData = mergedPlayerData.merge(combinedKills,left_on='match_id',right_on='match_id')
mergedPlayerData = mergedPlayerData.merge(combinedDeaths,left_on='match_id',right_on='match_id')

#Here we merge the two DataFrames together on the match_id columns, and make the match_id the index
merged = mergedPlayerData.merge(match,left_on='match_id',right_on='match_id')
merged = merged.set_index('match_id')
merged

dire_heros          radiant_heros  \
match_id                                                  
0          [106, 102, 46, 7, 73]   [86, 51, 83, 11, 67]   
1           [73, 22, 5, 67, 106]    [7, 82, 71, 39, 21]   
2            [38, 7, 10, 12, 85]   [51, 109, 9, 41, 27]   
3           [78, 19, 31, 40, 47]   [50, 44, 32, 26, 39]   
4         [101, 100, 22, 67, 21]    [8, 39, 55, 87, 69]   
...                          ...                    ...   
49995      [32, 7, 109, 35, 112]   [73, 86, 21, 20, 14]   
49996       [36, 1, 112, 60, 71]  [93, 74, 100, 32, 85]   
49997       [28, 102, 21, 9, 23]  [100, 68, 75, 39, 44]   
49998         [46, 7, 29, 44, 3]    [56, 50, 2, 72, 30]   
49999       [100, 9, 90, 73, 53]   [94, 19, 68, 35, 21]   

                     dire_heros_gpm          radiant_heros_gpm  \
match_id                                                         
0         [397, 303, 452, 189, 496]  [347, 494, 350, 599, 613]   
1         [600, 487, 488, 631, 585]  [218, 581, 339, 460, 365]   
2         [450, 376, 493, 389, 313]  [237, 322, 355, 328, 189]   
3         [459, 523, 323, 331, 500]  [290, 498, 454, 286, 453]   
4         [308, 235, 271, 418, 289]  [732, 707, 424, 348, 624]   
...                             ...                        ...   
49995     [468, 265, 551, 321, 254]  [619, 357, 429, 377, 456]   
49996     [343, 384, 142, 328, 255]  [540, 618, 459, 519, 476]   
49997     [290, 403, 319, 326, 323]  [463, 422, 480, 573, 566]   
49998     [264, 241, 321, 352, 286]  [523, 280, 458, 565, 306]   
49999     [468, 507, 371, 780, 520]  [430, 407, 414, 311, 406]   

                     dire_heros_xpm          radiant_heros_xpm  \
match_id                                                         
0         [524, 369, 517, 223, 456]  [362, 659, 385, 605, 762]   
1         [509, 517, 583, 755, 753]  [206, 756, 352, 544, 436]   
2         [567, 461, 535, 454, 363]  [249, 358, 425, 345, 229]   
3         [632, 630, 436, 302, 631]  [378, 619, 635, 351, 620]   
4         [309, 205, 313, 467, 272]  [683, 694, 372, 309, 541]   
...                             ...                        ...   
49995     [620, 309, 642, 349, 275]  [644, 426, 463, 411, 571]   
49996     [321, 356, 121, 331, 219]  [459, 613, 381, 445, 391]   
49997     [278, 539, 328, 378, 391]  [510, 469, 493, 697, 656]   
49998     [323, 256, 436, 389, 371]  [551, 387, 508, 636, 309]   
49999     [626, 607, 404, 703, 533]  [449, 611, 463, 398, 515]   

            dire_heros_kills radiant_heros_kills    dire_heros_deaths  \
match_id                                                                
0            [5, 4, 4, 1, 1]   [9, 13, 0, 8, 20]   [6, 13, 8, 14, 11]   
1         [2, 8, 14, 16, 10]    [3, 9, 5, 12, 6]     [12, 5, 8, 5, 7]   
2          [13, 5, 17, 7, 6]    [5, 6, 10, 0, 1]      [7, 6, 2, 2, 5]   
3         [13, 18, 7, 5, 20]  [4, 24, 17, 2, 17]  [8, 11, 16, 19, 11]   
4            [4, 3, 4, 4, 1]    [8, 18, 1, 1, 9]     [7, 9, 5, 5, 11]   
...                      ...                 ...                  ...   
49995      [24, 4, 15, 1, 1]    [7, 6, 8, 4, 18]    [11, 8, 9, 11, 6]   
49996        [3, 5, 0, 3, 5]    [3, 9, 7, 10, 9]     [8, 6, 11, 7, 7]   
49997       [5, 13, 5, 7, 1]  [16, 11, 6, 15, 6]  [11, 7, 13, 11, 12]   
49998        [1, 3, 4, 6, 5]    [10, 0, 7, 7, 4]      [6, 6, 3, 6, 8]   
49999      [16, 12, 5, 8, 6]    [2, 11, 6, 4, 7]      [9, 6, 3, 6, 6]   

          radiant_heros_deaths  radiant_win  duration  tower_status_radiant  \
match_id                                                                      
0              [3, 3, 4, 4, 3]         True      2375                  1982   
1          [4, 10, 13, 15, 11]        False      2582                     0   
2           [13, 11, 6, 9, 10]        False      2716                   256   
3         [13, 13, 11, 17, 12]        False      3085                     4   
4              [2, 4, 3, 4, 3]         True      1887                  2047   
...                        ...   

In [8]:
#The next step is to make the lists in the data into their own columns for the machine learning model
merged[["r1", "r2", "r3", "r4", "r5"]] = merged["radiant_heros"].tolist()
merged[["d1", "d2", "d3", "d4", "d5"]] = merged["dire_heros"].tolist()
merged[["r1_gpm", "r2_gpm", "r3_gpm", "r4_gpm", "r5_gpm"]] = merged["radiant_heros_gpm"].tolist()
merged[["d1_gpm", "d2_gpm", "d3_gpm", "d4_gpm", "d5_gpm"]] = merged["dire_heros_gpm"].tolist()
merged[["r1_xpm", "r2_xpm", "r3_xpm", "r4_xpm", "r5_xpm"]] = merged["radiant_heros_xpm"].tolist()
merged[["d1_xpm", "d2_xpm", "d3_xpm", "d4_xpm", "d5_xpm"]] = merged["dire_heros_xpm"].tolist()
merged[["r1_k", "r2_k", "r3_k", "r4_k", "r5_k"]] = merged["radiant_heros_kills"].tolist()
merged[["d1_k", "d2_k", "d3_k", "d4_k", "d5_k"]] = merged["dire_heros_kills"].tolist()
merged[["r1_d", "r2_d", "r3_d", "r4_d", "r5_d"]] = merged["radiant_heros_deaths"].tolist()
merged[["d1_d", "d2_d", "d3_d", "d4_d", "d5_d"]] = merged["dire_heros_deaths"].tolist()

merged = merged.drop(["radiant_heros","dire_heros"],axis=1)
merged = merged.drop(["radiant_heros_gpm","dire_heros_gpm"],axis=1)
merged = merged.drop(["radiant_heros_xpm","dire_heros_xpm"],axis=1)
merged = merged.drop(["radiant_heros_kills","dire_heros_kills"],axis=1)
merged = merged.drop(["radiant_heros_deaths","dire_heros_deaths"],axis=1)

merged

radiant_win  duration  tower_status_radiant  tower_status_dire  \
match_id                                                                   
0                True      2375                  1982                  4   
1               False      2582                     0               1846   
2               False      2716                   256               1972   
3               False      3085                     4               1924   
4                True      1887                  2047                  0   
...               ...       ...                   ...                ...   
49995            True      3025                  1824                  0   
49996            True      1451                  1982                384   
49997            True      2672                  1830                  0   
49998            True      2211                  1982                  4   
49999           False      2788                     0               2046   

          barracks_status_dire  barracks_status_radiant   r1   r2   r3  r4  \
match_id                                                                     
0                            3                       63   86   51   83  11   
1                           63                        0    7   82   71  39   
2                           63                       48   51  109    9  41   
3                           51                        3   50   44   32  26   
4                            0                       63    8   39   55  87   
...                        ...                      ...  ...  ...  ...  ..   
49995                        0                       63   73   86   21  20   
49996                       48                       63   93   74  100  32   
49997                        0                       63  100   68   75  39   
49998                        3                       63   56   50    2  72   
49999                       63                        0   94   19   68  35   

          ...  r1_d  r2_d  r3_d  r4_d  r5_d  d1_d  d2_d  d3_d  d4_d  d5_d  
match_id  ...                                                              
0         ...     3     3     4     4     3     6    13     8    14    11  
1         ...     4    10    13    15    11    12     5     8     5     7  
2         ...    13    11     6     9    10     7     6     2     2     5  
3         ...    13    13    11    17    12     8    11    16    19    11  
4         ...     2     4     3     4     3     7     9     5     5    11  
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
49995     ...     6    14    12     9     5    11     8     9    11     6  
49996     ...     3     2     3     5     3     8     6    11     7     7  
49997     ...    10     4     8     7     3    11     7    13    11    12  
49998     ...     5     7     5     2     4     6     6     3     6     8  
49999     ...    11    14     9    11     5     9     6     3     6     6  

[50000 rows x 56 columns]

In [9]:
#Here we start making the Machine Learning Model, we will start with a Random Forest Model

#Start by setting the prediction target
y = merged.radiant_win

#Next we set the features we are going to use
data_features = ["duration", "tower_status_radiant", "tower_status_dire", "barracks_status_dire", "barracks_status_radiant", "r1", "r2", "r3", "r4", "r5","d1", "d2", "d3", "d4", "d5","r1_gpm", "r2_gpm", "r3_gpm", "r4_gpm", "r5_gpm","d1_gpm", "d2_gpm", "d3_gpm", "d4_gpm", "d5_gpm","r1_xpm", "r2_xpm", "r3_xpm", "r4_xpm", "r5_xpm","d1_xpm", "d2_xpm", "d3_xpm", "d4_xpm", "d5_xpm","r1_k", "r2_k", "r3_k", "r4_k", "r5_k","d1_k", "d2_k", "d3_k", "d4_k", "d5_k","r1_d", "r2_d", "r3_d", "r4_d", "r5_d","d1_d", "d2_d", "d3_d", "d4_d", "d5_d"]
X = merged[data_features]

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics

#Here we split the data into training and validation data
train_X, val_X, train_y, val_y = train_test_split(X, y,random_state = 0)

#Make and fit a Random Forest Classifier Model using training data
forest_model = RandomForestClassifier(random_state=1)
forest_model.fit(train_X, train_y)

#Use the Model to predict the validation data and get a accuracy score based on that
preds = forest_model.predict(val_X)
print("Accuracy of the model is: ", metrics.accuracy_score(val_y, preds))

Accuracy of the model is:  0.99376


If we look at our accuracy score we see that we are correct 99.376% of the time, this looks great!

However there are a few things we must consider. We are using variables now that are available at the end of the match, for example the barrack status is going to largely correlated to the winning team.

So if we are planning on using this to predict a game before the game has finished our model will not work.

My goal is to be able to predict winning team for a live game. So to start we will need a way to get live match data.

We will use a website called Stratz, we are limited to a rate of 10,000 per day. “Powered by STRATZ”

In [10]:
#To start I will query for a match to use in predicting for our past machine learning model.

import requests

url = "https://api.stratz.com/graphql"
#This will be removed for privacy reasons but you can easily get one at: https://stratz.com/api
api_token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1laWQiOiJodHRwczovL3N0ZWFtY29tbXVuaXR5LmNvbS9vcGVuaWQvaWQvNzY1NjExOTgxMTgyMzcxODIiLCJ1bmlxdWVfbmFtZSI6IlBvd2VyIDwzIiwiU3ViamVjdCI6IjgyMmNiMTFmLTFjODEtNDI4Ni05YzFhLTI0NTQ5NDY0NzQwNiIsIlN0ZWFtSWQiOiIxNTc5NzE0NTQiLCJuYmYiOjE2NzgyMTE4NzksImV4cCI6MTcwOTc0Nzg3OSwiaWF0IjoxNjc4MjExODc5LCJpc3MiOiJodHRwczovL2FwaS5zdHJhdHouY29tIn0.XYUD159kVKKJCt0rAImWJZYh_gJC3fWJRSpeobcFZRk"
headers = {"Authorization": f"Bearer {api_token}"}

query = """
  {
    match(id: 7046446986) {
      didRadiantWin,
      durationSeconds,
      towerStatusRadiant,
      towerStatusDire,
      barracksStatusRadiant,
      barracksStatusDire,
      players{
          isRadiant,
          kills,
          deaths,
          assists,
          goldPerMinute,
          experiencePerMinute,
          hero{
              displayName,
          },
          steamAccount{
              name
          }
      }
    }
  }
"""

r = requests.post(url, json={"query":query}, headers=headers)
print(r.text)

{"data":{"match":{"didRadiantWin":false,"durationSeconds":2862,"towerStatusRadiant":4,"towerStatusDire":1982,"barracksStatusRadiant":3,"barracksStatusDire":63,"players":[{"isRadiant":true,"kills":2,"deaths":13,"assists":10,"goldPerMinute":300,"experiencePerMinute":408,"hero":{"displayName":"Tidehunter"},"steamAccount":{"name":"Noobraham Lincoln"}},{"isRadiant":true,"kills":7,"deaths":17,"assists":9,"goldPerMinute":307,"experiencePerMinute":440,"hero":{"displayName":"Witch Doctor"},"steamAccount":{"name":"..."}},{"isRadiant":true,"kills":7,"deaths":14,"assists":18,"goldPerMinute":497,"experiencePerMinute":544,"hero":{"displayName":"Bounty Hunter"},"steamAccount":{"name":"DIRTY GAY"}},{"isRadiant":true,"kills":7,"deaths":10,"assists":11,"goldPerMinute":350,"experiencePerMinute":498,"hero":{"displayName":"Winter Wyvern"},"steamAccount":{"name":"\u8CBB\u723E\u897F\u5C3C\u5967"}},{"isRadiant":true,"kills":9,"deaths":7,"assists":17,"goldPerMinute":572,"experiencePerMinute":720,"hero":{"displ

In [11]:
#Shows the data from the api in a prettier way
from pprint import pprint

pprint(r.json())

{'data': {'match': {'barracksStatusDire': 63,
                    'barracksStatusRadiant': 3,
                    'didRadiantWin': False,
                    'durationSeconds': 2862,
                    'players': [{'assists': 10,
                                 'deaths': 13,
                                 'experiencePerMinute': 408,
                                 'goldPerMinute': 300,
                                 'hero': {'displayName': 'Tidehunter'},
                                 'isRadiant': True,
                                 'kills': 2,
                                 'steamAccount': {'name': 'Noobraham Lincoln'}},
                                {'assists': 9,
                                 'deaths': 17,
                                 'experiencePerMinute': 440,
                                 'goldPerMinute': 307,
                                 'hero': {'displayName': 'Witch Doctor'},
                                 'isRadiant': True,
                    

I will start by using the website to get data for one game and see how our machine learning model predicts.

In [12]:
import json
data_features = ["duration", "tower_status_radiant", "tower_status_dire", "barracks_status_dire", "barracks_status_radiant", "r1", "r2", "r3", "r4", "r5","d1", "d2", "d3", "d4", "d5","r1_gpm", "r2_gpm", "r3_gpm", "r4_gpm", "r5_gpm","d1_gpm", "d2_gpm", "d3_gpm", "d4_gpm", "d5_gpm","r1_xpm", "r2_xpm", "r3_xpm", "r4_xpm", "r5_xpm","d1_xpm", "d2_xpm", "d3_xpm", "d4_xpm", "d5_xpm","r1_k", "r2_k", "r3_k", "r4_k", "r5_k","d1_k", "d2_k", "d3_k", "d4_k", "d5_k","r1_d", "r2_d", "r3_d", "r4_d", "r5_d","d1_d", "d2_d", "d3_d", "d4_d", "d5_d"]

data = json.loads(r.text)

#Get match data
duration = data["data"]["match"]["durationSeconds"]

#Get the kills,deaths and assists for each player
kills=[];
deaths= [];
assists= [];
for x in range(0,9):
    kills.append(data["data"]["match"]["players"][x]["kills"])
    deaths.append(data["data"]["match"]["players"][x]["deaths"])
    assists.append(data["data"]["match"]["players"][x]["assists"])

print(kills)
print(deaths)
print(assists)

#Make into format for machine learning data
gameData = pd.DataFrame([],columns=data_features)

print(gameData)
duration
#print(forest_model.predict(random_game))

[2, 7, 7, 7, 9, 17, 5, 8, 3]
[13, 17, 14, 10, 7, 4, 9, 8, 8]
[10, 9, 18, 11, 17, 18, 30, 27, 33]
Empty DataFrame
Columns: [duration, tower_status_radiant, tower_status_dire, barracks_status_dire, barracks_status_radiant, r1, r2, r3, r4, r5, d1, d2, d3, d4, d5, r1_gpm, r2_gpm, r3_gpm, r4_gpm, r5_gpm, d1_gpm, d2_gpm, d3_gpm, d4_gpm, d5_gpm, r1_xpm, r2_xpm, r3_xpm, r4_xpm, r5_xpm, d1_xpm, d2_xpm, d3_xpm, d4_xpm, d5_xpm, r1_k, r2_k, r3_k, r4_k, r5_k, d1_k, d2_k, d3_k, d4_k, d5_k, r1_d, r2_d, r3_d, r4_d, r5_d, d1_d, d2_d, d3_d, d4_d, d5_d]
Index: []

[0 rows x 55 columns]


2862

Using this website we can get events for the game. We can bucket these events into periods of time. For example we can check the winrate every 1min and make a prediction at that point.